new(scap): extend scap machine info, new agent info (new metrics 1/n)

incertum opened the pull request:
Signed-off-by: Melissa Kilby <melissa.kilby.oss@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

/area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

Migrating some lookups useful as identifiers and for new resource utilization metrics from Falco to libscap.

additional machine related info useful for:
* host attributions
* host kernel attributions
* identifiers for new resource utilization metrics logs

`scap_machine_info` seems a suitable place for new lookups. 

https://github.com/falcosecurity/falco/pull/2333
https://github.com/falcosecurity/falco/issues/2222


**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
new(scap): extend scap machine info, new agent info
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/880#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented on userspace/libscap/scap.c:
```diff
@@ -1464,6 +1465,57 @@ uint64_t scap_get_driver_schema_version(scap_t* handle)
+{
+	FILE* f;
+	// proc dir as seen from the calling process pid namespace
+	if((f = fopen("/proc/sys/kernel/bpf_stats_enabled", "r")))
```
This will be used to be able to activate `bpftool` like stats as we discussed here https://github.com/falcosecurity/falco/issues/2222#issuecomment-1312223538 and also seems a useful info to know about a machine as enabling bpf stats can add a slight performance overhead as stated in the docs https://manpages.ubuntu.com/manpages/focal/en/man8/bpftool-prog.8.html.

jasondellaluce commented on userspace/libscap/scap.c:
```diff
@@ -1464,6 +1465,57 @@ uint64_t scap_get_driver_schema_version(scap_t* handle)
+{
+	gethostname(handle->m_machine_info.hostname, sizeof(handle->m_machine_info.hostname) / sizeof(handle->m_machine_info.hostname[0]));
+	char *env_hostname;
+	env_hostname = getenv("FALCO_HOSTNAME");
```
The env var must not be `FALCO_` prefixed, but we'll need to make that a CMake overridable variable. Of course, for Falco as a consumer the name could be `FALCO_HOSTNAME`, but other libs consumers might want to have different namings.

Example of how we do this already for another env var:
- https://github.com/falcosecurity/libs/blob/b9591d68f44924f49bc9248638377b44a5328f56/userspace/libscap/scap.c#L1316
- https://github.com/falcosecurity/libs/blob/b9591d68f44924f49bc9248638377b44a5328f56/userspace/libscap/CMakeLists.txt#L44

jasondellaluce commented on userspace/libscap/scap.h:
```diff
@@ -948,6 +948,24 @@ uint64_t scap_get_driver_api_version(scap_t* handle);
 uint64_t scap_get_driver_schema_version(scap_t* handle);
 
+
+/**
```
is there a specific reason why are we exporting these functions in the libscap public API? Since we already have a getter to retrive machine info, and this info is populated in the handle at init time, I think these could just be internal helpers (scap_int.h maybe?). I would just avoid to further extend the API surface of libscap if not necessary, specially now that we're doing such a great job in its polishing process.

gnosek replied:
Agreed. Also, maybe it's a good time to factor out filling machine_info to a separate function (used for live/udig/nodriver engines)?

jasondellaluce commented on userspace/libscap/scap.h:
```diff
@@ -324,9 +324,9 @@ typedef struct _scap_machine_info
-	uint64_t reserved2; ///< reserved for future use
-	uint64_t reserved3; ///< reserved for future use
-	uint64_t reserved4; ///< reserved for future use
+	uint64_t self_pid_start_ts; ///< Agent start ts in nanoseconds (epoch)
```
Sounds good, but I miss the context on why these were reserved. Maybe because of padding and backward compatibility with something?

My only concern, and broader question, is: will we be able to further grow this struct in the future, when we need more info?

gnosek replied:
AFAIK it's for future extensibility (so that we don't need another hack while reading scap files). Note that this patch breaks compatibility already (the old/new sizes are very much different) so we need changes to savefile code in this PR.

Personally, I'd rather:
* add the string field _after_ the currently reserved fields (let's keep the reserved ones for future integer vars)
* use an `uint64_t flags` instead of a bool and `#define` a numeric constant for the one bit we need so far (BTW we can't use `bool` anyway since its size [isn't guaranteed to be the same](https://stackoverflow.com/a/4897859/922138) on all platforms and we save this struct directly to disk)

(deep in my patch queue there's another patch that adds stuff to machine_info; it needs just a few bits, like 3 or 4 I think, so maybe it could piggyback on the flags field without adding yet another one and using up a pretty scarce resource)

jasondellaluce reviewed (commented):
Looks good so far! Left you some points.

incertum commented:
Thanks for the great initial feedback @gnosek and @jasondellaluce - I see the complications here, let's change it. I wasn't sure anyways where to put this data and how to best expose it with the bigger scap cleanup picture in mind. I like all of your suggestions, @gnosek I'll coordinate with you.

incertum commented:
Addressed initial feedback. Small follow up question if I may: Given that scap_machine_info fields unfortunately are a scarce resource plus the type limitations is it ok to keep `self_pid_ts_epoch` in there? Because would be open to just add it to libsinsp as simple getter? Also `uname_r` got purged, we can keep it in Falco right where we generate the new resource utilization logs.

jasondellaluce commented on userspace/libscap/scap.c:
```diff
@@ -1513,6 +1513,51 @@ uint64_t scap_get_driver_schema_version(scap_t* handle)
+void scap_get_bpf_stats_enabled(scap_t* handle)
+{
+	FILE* f;
+	// proc dir as seen from the calling process pid namespace
```
```suggestion
	handle->m_machine_info.flags &= ~PPM_BPF_STATS_ENABLED;
	// proc dir as seen from the calling process pid namespace
```
Making sure we clear out the flag bit.

jasondellaluce reviewed (commented):
Left some more comments! Also, I'd put emphasis on the discussion at: https://github.com/falcosecurity/libs/pull/880/files#r1104380474

incertum commented:
Addressed reviewers comments including the request to cleanup machine_info retrieval via a clean function.

Would adding the new proposed `scap_agent_info` work? Could be that it may come in handy in the future for extensions, especially because of the limitations of `scap_machine_info` because of the scap files.

Plus because of the boot time hiccup decided to retrieve start_time from the agent stat file in order to base CPU usage calculations on that without needing to rely on procfs time stamps, but I kept that metric too so we can report a constant start timestamp of the agent as it should work on most systems.

Andreagit97 commented on userspace/libsinsp/sinsp.cpp:
```diff
@@ -1846,6 +1846,11 @@ const scap_machine_info* sinsp::get_machine_info()
 
+const scap_agent_info* sinsp::get_agent_info()
+{
+	return m_agent_info;
```
right now we are not populating `m_agent_info` is this correct?

jasondellaluce replied:
+1

Andreagit97 commented on userspace/libscap/scap.h:
```diff
@@ -1002,6 +1021,31 @@ uint64_t scap_get_driver_api_version(scap_t* handle);
+ * where the hostname can be equivalent to the Kubernetes pod name.
+ * Customizable over cmake setup SCAP_HOSTNAME_ENV_VAR.
+ */
+void scap_gethostname(scap_t* handle);
+
+/**
+ * Retrieve agent info.
+ */
+void scap_retrieve_agent_info(scap_t* handle);
+
+/**
+ * Retrieve machine info.
+ */
+void scap_retrieve_machine_info(scap_t* handle, uint64_t boot_time);
+
+/**
+ * Check is kernel.bpf_stats_enabled is set.
+ */
+void scap_get_bpf_stats_enabled(scap_t* handle);
```
As @jasondellaluce correctly said these seem more like internal scap APIs rather than public ones, probably they should live in `scap-int.h` and we could have a dedicated `.c` file for the implementation (similar to what we do with `scap_userlists.c`

incertum replied:
@jasondellaluce and @Andreagit97 will do.

incertum replied:
oops almost forgot about the request to create a dedicated `.c` file, just pushed it, see last commit @Andreagit97.

jasondellaluce commented on userspace/libscap/scap.h:
```diff
@@ -324,11 +324,21 @@ typedef struct _scap_machine_info
+typedef struct _scap_agent_info
+{
+	uint64_t start_ts_epoch; ///< Agent start timestamp, stat /proc/self/cmdline approach, unit: epoch in nanoseconds
+	double start_time; ///< /proc/self/stat start_time divided by HZ, unit: seconds
```
Since the unit is seconds, I think we could make this `uint64_t` too and round-up the division.

incertum replied:
@jasondellaluce a bit worried about precision loss in general whenever we do a division which up until now we are not doing much in the project. For example see https://gitlab.com/procps-ng/procps/-/blob/master/library/pids.c#L286. Would it be ok to keep it as double as I would want to keep elapsed time of the process as double as well, because that will be used for division in CPU usage calculation at a later point in time.

jasondellaluce reviewed (commented):
Looks good Melissa, thanks for addressing all the concerns. I left few comments, but I think my only remaining big point is to avoid adding all those new functions in `scap.h` and just keep there `const scap_agent_info* scap_get_agent_info(scap_t* handle);`. You can either add those functions to `scap-int.h` or have them defined as static in `scap.c`, but my point is simply to avoid having those as part of the exported API unless we have a concrete use case.

incertum commented on userspace/libscap/scap.h:
```diff
@@ -1002,6 +1021,16 @@ uint64_t scap_get_driver_api_version(scap_t* handle);
+ * where the hostname can be equivalent to the Kubernetes pod name.
+ * Customizable over cmake setup SCAP_HOSTNAME_ENV_VAR.
+ */
+void scap_gethostname(scap_t* handle);
```
Done, also had to rebase one more time.

poiana commented:
LGTM label has been added.  <details>Git tree hash: b36581820ae161c291a816f4a29e25ef1fa06605</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/880#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/libs/pull/880#pullrequestreview-1361376715" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [incertum,jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 reviewed (approved):
/approve
